### PR TITLE
analyzer: Reconnect stream consumer after some time without messages

### PR DESCRIPTION
**Pull request by victorges**
This is to improve reliability of the analyzer, by detecting if we get no messages from rabbimtq
and restarting the consumer manually to fix it if so. This is because we've noticed that the amqp
stream consumer lib fails to detect when RabbitMQ clusters are disconnected and just stays
there for a long time without receiving any messages until we manually restart the analyzers.

This fixes that by reconnecting the stream consumer every 3 minutes if no messages are received.

### CODE REVIEW

**Comment by tqian1** [comment]
Couple q's

Any ideas on how come stream consumer lib fails to detect disconnected cluster? 

Does it make sense to check the connection/messages received as part of the readiness check and restart the entire pod? What causes the lib to disconnect from the cluster? Does the reconnect always work when its triggered?

**Comment by victorges** [comment]
@tqian1 
> Any ideas on how come stream consumer lib fails to detect disconnected cluster?
> What causes the lib to disconnect from the cluster? 

Not really sure... It apparently happens even when the RabbitMQ cluster apparently didn't have any issues. There is no specific log from the library here, it just stops trying to reconnect/receiving messages, so not sure... Ideally we would create an issue on the client library (https://github.com/rabbitmq/rabbitmq-stream-go-client) but I don't think we have enough evidence for them to be able to reproduce/fix it anyway...

> Does it make sense to check the connection/messages received as part of the readiness check and restart the entire pod? 

I don't think that's necessary since restarts are accounted for in the state of the kubernetes pod while I know/hope that just recreating the consumer will already do it. I prefer to leave panic/restarts to errors that we absolutely can do nothing about and needs a full process restart. In this case, the lib gets to some kind of deadlock and recreating the consumer should work. If it doesn't, we can revert this change anyway.

We also already have a [logic to check for RabbitMQ connectivity on the readiness probe](https://github.com/livepeer/livepeer-data/blob/694314cac21fc7e5bba0c54f9257d882cb74d3a8/health/core.go#L85), but that actually creates a new client and sends a dummy command on it. That won't restart the pod though, only stop sending requests to it ([k8s docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)), and I think that's enough.

> Does the reconnect always work when its triggered?

Not sure yet. It's a hard to reproduce problem so I was just thinking of getting this to production and [checking the new logs](https://eu-metrics-monitoring.livepeer.live/grafana/explore?orgId=1&left=%5B%22now-2d%22,%22now%22,%22Loki%22,%7B%22exemplar%22:true,%22expr%22:%22%7Bapp%3D%5C%22staging-livepeer-analyzer%5C%22%7D%20%7C%3D%20%5C%22Reconnecting%20after%5C%22%22,%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22P8E80F9AEF21F6940%22%7D%7D%5D) (seems like we have some in staging) later to check if it had the desired effect. So the important thing here is to review that it won't have a bad impact on the system in production when everything is going ok.

**Comment by tqian1** [comment]
`We also already have a [logic to check for RabbitMQ connectivity on the readiness probe](https://github.com/livepeer/livepeer-data/blob/694314cac21fc7e5bba0c54f9257d882cb74d3a8/health/core.go#L85), but that actually creates a new client and sends a dummy command on it. That won't restart the pod though, only stop sending requests to it ([k8s docs](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/)), and I think that's enough.`

Got it so the newly created client wouldn't detect the problem - that readiness probe would only fail if the RabbitMQ cluster was actually down/unresponsive and the new client was unable to connect if i understand correctly.

File: pkg/event/stream_consumer.go [hunk 1]
```diff
@@ -159,9 +159,13 @@ func newReconnectingConsumer(ctx context.Context, streamName string, connect con
 	}
 
 	var (
+		reconnect = func() (*stream.Consumer, bool) {
+			return ensureConnect(ctx, streamName, connect, stream.ConsumerContext{Consumer: consumer}, msgHandler)
+		}
 		done                = make(chan struct{})
 		reconnectTimer      <-chan time.Time
 		resetReconnectTimer = func() {}
+		ok                  bool
```
Comment (tqian1) [inline pkg/event/stream_consumer.go:168]: nit*
Could be more descriptive, ie. `connected`

File: pkg/event/stream_consumer.go [hunk 2]
```diff
@@ -184,8 +188,7 @@ func newReconnectingConsumer(ctx context.Context, streamName string, connect con
 				return
 			case ev := <-consumer.NotifyClose():
 				glog.Errorf("Stream consumer closed, reconnecting. consumer=%q, stream=%q, reason=%q", ev.Name, ev.StreamName, ev.Reason)
-				consumer = ensureConnect(ctx, streamName, connect, stream.ConsumerContext{Consumer: consumer}, msgHandler)
-				if consumer == nil {
+				if consumer, ok = reconnect(); !ok {
```
Comment (tqian1) [inline pkg/event/stream_consumer.go:191]: Makes sense.

**Review by tqian1 — APPROVED** [review]
LGTM, didn't find any problems with the reconnection logic ie. reconnecting unnecessarily or in a loop

The syntax/logic around channels were advanced for me

**Comment by victorges** [comment]
I'm gonna close this for now since it seems like the change will be a little more complex than expected, and we haven't observed a lot of this since the latest improvements to RabbitMQ and the clusters infrastructure.

If it seems to become useful again in the future, can always pick it back up.

**Pull request closed without merging**